### PR TITLE
Benchmark framework extensions

### PR DESCRIPTION
**Pull request by AdamHarries**
This PR adds functionality to the benchmarks in the form of performance reporting based on both overall, and event timings.

### CODE REVIEW

File: bench/common/blas_benchmark.hpp [hunk 1]
```diff
@@ -75,54 +75,219 @@ const inline std::string& type_string<double>() {
...
+        break;
+      case CL_OUT_OF_RESOURCES:
+        std::cerr << "cl_out_of_resources in profiling call" << std::endl;
+        break;
+      case CL_OUT_OF_HOST_MEMORY:
+        std::cerr << "cl_out_of_host_memory in profiling call" << std::endl;
+        break;
+    }
+    // If we've reached this point, something's gone wrong - set the error flag
+    all_ok = false;
```
Comment (Ruyk) [inline bench/common/blas_benchmark.hpp]: why not use the default of the switch for this?

File: bench/common/blas_benchmark.hpp [hunk 2]
```diff
@@ -75,54 +75,219 @@ const inline std::string& type_string<double>() {
...
+  check_call(clGetEventProfilingInfo(e._cl(), CL_PROFILING_COMMAND_START,
+                                     sizeof(cl_ulong), &start_time, NULL));
+  check_call(clGetEventProfilingInfo(e._cl(), CL_PROFILING_COMMAND_START,
+                                     sizeof(cl_ulong), &end_time, NULL));
+  e.release();
+  // return the delta
+  if (all_ok) {
+    return (end_time - start_time);
+  } else {
+    return (cl_ulong)(-1);
```
Comment (Ruyk) [inline bench/common/blas_benchmark.hpp]: cl_ulong is unsigned long, but you return -1?

File: bench/common/blas_benchmark.hpp [hunk 3]
```diff
@@ -75,54 +75,219 @@ const inline std::string& type_string<double>() {
...
+  void output_data(const std::string& name, int no_reps,
+                   output_type output = output_type::STDOUT) {
+    // Calculate GFlop/s overall, and for the kernel.
+    // As it turns out, Flop/Ns (i.e. floating point operations per nanosecond)
+    // is the same as GFlops/S (i.e. billion floating point operations per
+    // second), so we don't need to do any duration casts between time
+    // intervals, or between flops and gigaflops - we can just compute Flop/Ns
+    // directly.
+    auto _mean_overall_flops = (double)(_n_fl_ops) / _mean_overall_time.count();
+    auto _mean_kernel_flops = (double)(_n_fl_ops) / _mean_kernel_time.count();
```
Comment (Ruyk) [inline bench/common/blas_benchmark.hpp]: this should be static cast instead of C cast

File: bench/syclblas/syclblas_benchmark_level1.cpp [hunk 4]
```diff
@@ -45,10 +45,11 @@ BENCHMARK(scal, syclblas_level_1) {
 
   auto in = ex.template allocate<ScalarT>(size);
   ex.copy_to_device(v1.data(), in, size);
-  benchmark<>::flops_units_t flops =
-      benchmark<>::measure(reps, size * 1, [&]() {
+  benchmark<>::datapoint_t flops =
+      benchmark<>::measure(reps, size * 1, [&]() -> cl::sycl::event {
```
Comment (Ruyk) [inline bench/syclblas/syclblas_benchmark_level1.cpp]: why the explicit return type?

File: bench/common/blas_benchmark.hpp (hunk 1 above)
Comment (AdamHarries) [reply bench/common/blas_benchmark.hpp]: good point!

File: bench/common/blas_benchmark.hpp (hunk 2 above)
Comment (AdamHarries) [reply bench/common/blas_benchmark.hpp]: Yes - if there's a problem with the event, we don't really want to fail (this is not a test, after all), but we want to produce the largest number we can. It occurs that this should probably be the equivalent of UINT_MAX for cl_ulong

File: bench/syclblas/syclblas_benchmark_level1.cpp (hunk 4 above)
Comment (AdamHarries) [reply bench/syclblas/syclblas_benchmark_level1.cpp]: It's not required, but it produces errors more clearly if we forget to return an event.

File: bench/clblast/clblast_benchmark_level1.cpp [hunk 5]
```diff
@@ -71,13 +70,12 @@ BENCHMARK(axpy, clblast_level_1) {
...
-        clblast::Axpy<ScalarT>(size, alpha, buf1.dev(), 0, 1, buf2.dev(), 0, 1,
-                               (*ex)._queue(), &event._cl());
-        event.wait();
-        event.release();
-      });
+  benchmark<>::datapoint_t flops = benchmark<>::measure(reps, size * 2, [&]() {
+    clblast::Axpy<ScalarT>(size, alpha, buf1.dev(), 0, 1, buf2.dev(), 0, 1,
+                           (*ex)._queue(), &event._cl());
+    event.wait();
+    return event;
```
[the remark this answers is not in the document]
Comment (AdamHarries) [reply bench/clblast/clblast_benchmark_level1.cpp]: the event.release has been moved to the measuring code outside the lambda - inside timef, in blas_benchmark.cpp

**Pull request merged**